utils/zerovec/src/map2k/mod.rs

```diff
@@ -0,0 +1,687 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+//! See [`ZeroMap2k`] for details.
+
+use crate::ule::AsULE;
+use crate::ZeroVec;
+use core::cmp::Ordering;
+use core::convert::TryFrom;
+use core::fmt;
+use core::ops::Range;
+
+mod borrowed;
+#[cfg(feature = "serde")]
+mod serde;
+
+use crate::map::ZeroMapKV;
+use crate::map::{MutableZeroVecLike, ZeroVecLike};
+pub use borrowed::ZeroMap2kBorrowed;
+
+/// A zero-copy, two-dimensional map datastructure .
+///
+/// This is an extension of [`ZeroMap`] that supports two layers of keys. For example,
+/// to map a pair of an integer and a string to a buffer, you can write:
+///
+/// ```no_run
+/// # use zerovec::ZeroMap2k;
+/// let _: ZeroMap2k<u32, str, [u8]> = unimplemented!();
+/// ```
+///
+/// Internally, `ZeroMap2k` stores four zero-copy vectors, one for each type argument plus
+/// one more to match between the two vectors of keys.
+///
+/// # Examples
+///
+/// ```
+/// use zerovec::ZeroMap2k;
+///
+/// // Example byte buffer representing the map { 1: {2: "three" } }
+/// let BINCODE_BYTES: &[u8; 53] = &[
+///     2, 0, 0, 0, 0, 0, 0, 0, 1, 0, 4, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 0, 0, 0, 0, 0,
+///     0, 0, 2, 0, 13, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 116, 104, 114, 101, 101
+/// ];
+///
+/// // Deserializing to ZeroMap requires no heap allocations.
+/// let zero_map: ZeroMap2k<u16, u16, str> = bincode::deserialize(BINCODE_BYTES)
+///     .expect("Should deserialize successfully");
+/// assert_eq!(zero_map.get(&1, &2), Some("three"));
+/// ```
+///
+/// [`VarZeroVec`]: crate::VarZeroVec
+/// [`ZeroMap`]: crate::ZeroMap
+pub struct ZeroMap2k<'a, K0, K1, V>
```

Comment:
thought: not a fan of the name, "2k" makes me think "2000". `ZeroMapDoubleKey` perhaps? But that's even longer. Maybe `DoubleZeroMap`.

Not against 2k though; fine with it landing with that name, just hoping for something better

Reply:
I got `2k` from "two keys", and it seems like a harmless analogy to 2000.  The other name I was considering was `ZeroMap2d` for two-dimensional.

Reply:
Yeah I think I prefer 2d.